File: ga/gpt3_5_result_2024_6/result/queue_gpt/25_queue.c
use std::ptr;
use std::process;

type ElementType = i32;

struct QueueRecord {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Box<[ElementType]>,
}

struct Queue {
    queue_record: QueueRecord,
}

const MIN_QUEUE_SIZE: usize = 5;

impl Queue {
    fn is_empty(&self) -> bool {
        self.queue_record.size == 0
    }

    fn is_full(&self) -> bool {
        self.queue_record.size == self.queue_record.capacity
    }

    fn create_queue(max_elements: usize) -> Queue {
        if max_elements < MIN_QUEUE_SIZE {
            eprintln!("Queue size is too small");
            process::exit(1);
        }

        let array = vec![0; max_elements].into_boxed_slice();

        Queue {
            queue_record: QueueRecord {
                capacity: max_elements,
                front: 1,
                rear: 0,
                size: 0,
                array,
            },
        }
    }

    fn make_empty(&mut self) {
        self.queue_record.size = 0;
        self.queue_record.front = 1;
        self.queue_record.rear = 0;
    }

    fn enqueue(&mut self, x: ElementType) {
        if self.is_full() {
            eprintln!("Full queue");
            process::exit(1);
        } else {
            let rear = (self.queue_record.rear + 1) % self.queue_record.capacity;
            self.queue_record.size += 1;
            self.queue_record.rear = rear;
            self.queue_record.array[self.queue_record.rear] = x;
        }
    }

    fn front(&self) -> ElementType {
        if !self.is_empty() {
            self.queue_record.array[self.queue_record.front]
        } else {
            eprintln!("Empty queue");
            process::exit(1);
        }
    }

    fn dequeue(&mut self) {
        if self.is_empty() {
            eprintln!("Empty queue");
            process::exit(1);
        } else {
            self.queue_record.size -= 1;
            self.queue_record.front = (self.queue_record.front + 1) % self.queue_record.capacity;
        }
    }

    fn front_and_dequeue(&mut self) -> ElementType {
        let x: ElementType;
        
        if self.is_empty() {
            eprintln!("Empty queue");
            process::exit(1);
        } else {
            self.queue_record.size -= 1;
            x = self.queue_record.array[self.queue_record.front];
            self.queue_record.front = (self.queue_record.front + 1) % self.queue_record.capacity;
        }
        
        x
    }

    fn dispose_queue(&mut self) {
        self.queue_record.size = 0;
        self.queue_record.front = 0;
        self.queue_record.rear = self.queue_record.capacity - 1;
        self.queue_record.array = Box::new([]);
    }
}

fn main() {
    let mut q = Queue::create_queue(12);

    for i in 0..10 {
        q.enqueue(i);
    }

    while !q.is_empty() {
        println!("{}", q.front());
        q.dequeue();
    }

    for i in 0..10 {
        q.enqueue(i);
    }

    while !q.is_empty() {
        println!("{}", q.front());
        q.dequeue();
    }

    q.dispose_queue();
}

File: ga/4omini_result_2024_7/result/queue_mini/25_queue.rs
use std::io::{self, Write};

#[derive(Debug)]
pub struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec<i32>,
}

impl Queue {
    // Create a new queue with the specified maximum capacity
    pub fn new(max_elements: usize) -> Queue {
        if max_elements < 5 {
            panic!("Queue size is too small");
        }
        Queue {
            capacity: max_elements,
            front: 0,
            rear: max_elements - 1,
            size: 0,
            array: vec![0; max_elements],
        }
    }

    // Check if the queue is empty
    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    // Check if the queue is full
    pub fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    // Enqueue an element to the queue
    pub fn enqueue(&mut self, x: i32) {
        if self.is_full() {
            panic!("Full queue");
        } else {
            self.rear = (self.rear + 1) % self.capacity;
            self.array[self.rear] = x;
            self.size += 1;
        }
    }

    // Get the front element without dequeuing
    pub fn front(&self) -> i32 {
        if self.is_empty() {
            panic!("Empty queue");
        }
        self.array[self.front]
    }

    // Dequeue an element from the queue
    pub fn dequeue(&mut self) {
        if self.is_empty() {
            panic!("Empty queue");
        } else {
            self.front = (self.front + 1) % self.capacity;
            self.size -= 1;
        }
    }

    // Get the front element and dequeue it
    pub fn front_and_dequeue(&mut self) -> i32 {
        let x = self.front();
        self.dequeue();
        x
    }
}

fn main() {
    let mut q = Queue::new(12);

    // Enqueue integers 0 to 9
    for i in 0..10 {
        q.enqueue(i);
    }

    // Print and dequeue until empty
    while !q.is_empty() {
        println!("{}", q.front());
        q.dequeue();
    }

    // Enqueue integers 0 to 9 again
    for i in 0..10 {
        q.enqueue(i);
    }

    // Print and dequeue until empty
    while !q.is_empty() {
        println!("{}", q.front());
        q.dequeue();
    }
}